core/backend/src/BriefcaseManager.ts

```diff
@@ -432,6 +444,59 @@ export class BriefcaseManager {
     IModelJsFs.removeSync(changesetFile.pathname);
   }
 
+  /** @internal */
+  public static async revertTimelineChanges(db: IModelDb, arg: RevertChangesArgs): Promise<void> {
+    if (!db.isOpen || db[_nativeDb].isReadonly()) // don't use db.isReadonly - we reopen the file writable just for this operation but db.isReadonly is still true
+      throw new IModelError(ChangeSetStatus.ApplyError, "Briefcase must be open ReadWrite to process change sets");
+
+    let currentIndex = db.changeset.index;
+    if (currentIndex === undefined)
+      currentIndex = (await IModelHost.hubAccess.queryChangeset({ accessToken: arg.accessToken, iModelId: db.iModelId, changeset: { id: db.changeset.id } })).index;
+
+    if (!arg.toIndex) {
+      throw new IModelError(ChangeSetStatus.ApplyError, "toIndex must be specified to revert changesets");
+    }
+    if (arg.toIndex > currentIndex) {
+      throw new IModelError(ChangeSetStatus.ApplyError, "toIndex must be less than the current index to revert changesets");
+    }
+    if (!db.holdsSchemaLock) {
+      throw new IModelError(ChangeSetStatus.ApplyError, "Cannot revert timeline changesets without holding a schema lock");
+    }
+
+    // Download change sets
+    const changesets = await IModelHost.hubAccess.downloadChangesets({
+      accessToken: arg.accessToken,
+      iModelId: db.iModelId,
+      range: { first: arg.toIndex, end: currentIndex },
```

Comment:
Why not pass `end: undefined` to make sure that all change sets are retrieved, other option could be to lock schema before pulling latest changes:
![image](https://github.com/user-attachments/assets/89cb8c67-0e4d-43ca-9a2e-07a9ddb2a8a4)

Reply:
I had to push because in schema sync we can get a schema changeset. I have remove this code and now we never do schemaSync.pull and so there will be no local changeset after pull.

Reply:
This case is now also covered by the test.